### PR TITLE
fix: add condition in method ensure_kubeconfig!

### PR DESCRIPTION
**Pull request by barmull**
## Description
 - add one more elsif condition in case variable KUBECONFIG is not set (empty)
  - change error messages

## Issues:
Refs: #985 

## How has this been tested:
 - [ ] Covered by existing integration testing
 - [ ] Added integration testing to cover
 - [ ] Verified all A/C passes
     * [ ] develop
     * [ ] master
     * [ ] tag/other branch
 - [ ] Test environment
    * [ ] Shared Packet K8s cluster
    * [ ] New Packet K8s cluster
    * [X] Kind cluster
 - [ ] Have not tested

## Types of changes:
 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update

## Checklist:
**Documentation**
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] No updates required.

**Code Review**
- [ ] Does the test handle fatal exceptions, ie. rescue block

**Issue**
- [ ] Tasks in issue are checked off

### CODE REVIEW

File: src/tasks/utils/utils.cr [hunk 1]
```diff
@@ -44,9 +44,15 @@ def ensure_kubeconfig!
...
     puts "KUBECONFIG is set as #{ENV["KUBECONFIG"]}.".colorize(:green)
+  elsif !ENV.has_key?("KUBECONFIG")
+    ENV["KUBECONFIG"] = kubeconfig_path
+    error_message = "KUBECONFIG is not set (variable is empty) and default path #{ENV["KUBECONFIG"]} does not exist. Please set a KUBECONFIG to existing config file, i.p 'export KUBECONFIG=path-to-your-kubeconfig'"
+    puts error_message.colorize(:red)
+    raise error_message
   else
-    puts "KUBECONFIG is not set. Please set a KUBECONFIG, i.p 'export KUBECONFIG=path-to-your-kubeconfig'".colorize(:red)
-    raise "KUBECONFIG is not set. Please set a KUBECONFIG, i.p 'export KUBECONFIG=path-to-your-kubeconfig'"
+    error_message = "KUBECONFIG path does not exist. Please set a KUBECONFIG to existing config file, i.p 'export KUBECONFIG=path-to-your-kubeconfig'"
```
Comment (martin-mat) [inline src/tasks/utils/utils.cr:53]: error_message = "KUBECONFIG is set to #{ENV["KUBECONFIG"} path and it does not exist. Please set a KUBECONFIG to an existing config file, i.e. 'export KUBECONFIG=path-to-your-kubeconfig'"

**Review by martin-mat — CHANGES_REQUESTED** [review]

**Comment by martin-mat** [comment]
lgtm

**Review by taylor — COMMENTED** [review]
lgtm

**Comment by taylor** [comment]
⬆️  @wavell @denverwilliams take a look and merge if you 👍🏻

**Review by martin-mat — APPROVED** [review]
lgtm

**Review by agentpoyo — CHANGES_REQUESTED** [review]
@barmull I tested this, seems when KUBECONFIG is not set, it's defaulting to the existing output message `Could not create cnf-testsuite namespace on the Kubernetes cluster`... see below where I unset KUBECONFIG:

```
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ echo $KUBECONFIG
/home/pair/workspace/drew/blah
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ unset KUBECONFIG
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ echo $KUBECONFIG

pair@cnfdev4:~/workspace/drew/testsuite-barmull$ ./cnf-testsuite setup
CNF TestSuite version: HEAD-2024-04-20-204358-4cfa4202
Successfully created directories for cnf-testsuite
Global helm found. Version: v3.11.1
No Local helm version found
Global kubectl found. Version: 1.26
Global kubectl client version could not be checked for compatibility with server. (Server not configured?)
No Local kubectl version found
Global git found. Version: 2.34.1
No Local git version found
All prerequisites found.
KUBECONFIG is set as /home/pair/.kube/config.
Could not create cnf-testsuite namespace on the Kubernetes cluster
Setup complete
```

This is likely due to `KUBECONFIG is set as /home/pair/.kube/config.`

When I remove the local .kube/config for my user, the result output is this:
```
$ ./cnf-testsuite setup
CNF TestSuite version: HEAD-2024-04-20-204914-4cfa4202
Successfully created directories for cnf-testsuite
Global helm found. Version: v3.11.1
No Local helm version found
Global kubectl found. Version: 1.26
Global kubectl client version could not be checked for compatibility with server. (Server not configured?)
No Local kubectl version found
Global git found. Version: 2.34.1
No Local git version found
All prerequisites found.
KUBECONFIG is not set and default path /home/pair/.kube/config does not exist. Please set KUBECONFIG to an existing config file, i.e. 'export KUBECONFIG=path-to-your-kubeconfig'
src/tasks/utils/utils.cr:52:45 in 'ensure_kubeconfig!'
src/tasks/setup.cr:13:3 in '->'
lib/sam/src/sam/task.cr:54:39 in 'call'
lib/sam/src/sam/execution.cr:20:7 in 'invoke'
lib/sam/src/sam/task.cr:44:29 in 'call'
lib/sam/src/sam/execution.cr:20:7 in 'invoke'
lib/sam/src/sam.cr:35:5 in 'invoke'
lib/sam/src/sam.cr:53:7 in 'process_tasks'
src/cnf-testsuite.cr:132:3 in '__crystal_main'
/home/pair/.crenv/versions/1.6.2/share/crystal/src/crystal/main.cr:115:5 in 'main_user_code'
/home/pair/.crenv/versions/1.6.2/share/crystal/src/crystal/main.cr:101:7 in 'main'
/home/pair/.crenv/versions/1.6.2/share/crystal/src/crystal/main.cr:127:3 in 'main'
/lib/x86_64-linux-gnu/libc.so.6 in '??'
/lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
./cnf-testsuite in '_start'
???
KUBECONFIG is not set and default path /home/pair/.kube/config does not exist. Please set KUBECONFIG to an existing config file, i.e. 'export KUBECONFIG=path-to-your-kubeconfig'
```

Here are the outputs when I export KUBECONFIG to a non working config and then a working config for KUBECONFIG:

```
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ export KUBECONFIG=/home/pair/workspace/drew/3node.kubeconfig
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ ./cnf-testsuite setup
CNF TestSuite version: HEAD-2024-04-20-205225-4cfa4202
Successfully created directories for cnf-testsuite
Global helm found. Version: v3.11.1
No Local helm version found
Global kubectl found. Version: 1.26
No Local kubectl version found
Global git found. Version: 2.34.1
No Local git version found
All prerequisites found.
KUBECONFIG is already set.
cnf-testsuite namespace already exists on the Kubernetes cluster
Setup complete
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ export KUBECONFIG=/home/pair/workspace/drew/blah
pair@cnfdev4:~/workspace/drew/testsuite-barmull$ ./cnf-testsuite setup
CNF TestSuite version: HEAD-2024-04-20-205248-4cfa4202
Successfully created directories for cnf-testsuite
Global helm found. Version: v3.11.1
No Local helm version found
Global kubectl found. Version: 1.26
Global kubectl client version could not be checked for compatibility with server. (Server not configured?)
No Local kubectl version found
Global git found. Version: 2.34.1
No Local git version found
All prerequisites found.
KUBECONFIG is set to /home/pair/workspace/drew/blah path and it does not exist. Please set KUBECONFIG to an existing config file, i.e. 'export KUBECONFIG=path-to-your-kubeconfig'
src/tasks/utils/utils.cr:54:5 in 'ensure_kubeconfig!'
src/tasks/setup.cr:13:3 in '->'
lib/sam/src/sam/task.cr:54:39 in 'call'
lib/sam/src/sam/execution.cr:20:7 in 'invoke'
lib/sam/src/sam/task.cr:44:29 in 'call'
lib/sam/src/sam/execution.cr:20:7 in 'invoke'
lib/sam/src/sam.cr:35:5 in 'invoke'
lib/sam/src/sam.cr:53:7 in 'process_tasks'
src/cnf-testsuite.cr:132:3 in '__crystal_main'
/home/pair/.crenv/versions/1.6.2/share/crystal/src/crystal/main.cr:115:5 in 'main_user_code'
/home/pair/.crenv/versions/1.6.2/share/crystal/src/crystal/main.cr:101:7 in 'main'
/home/pair/.crenv/versions/1.6.2/share/crystal/src/crystal/main.cr:127:3 in 'main'
/lib/x86_64-linux-gnu/libc.so.6 in '??'
/lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
./cnf-testsuite in '_start'
???
KUBECONFIG is set to /home/pair/workspace/drew/blah path and it does not exist. Please set KUBECONFIG to an existing config file, i.e. 'export KUBECONFIG=path-to-your-kubeconfig'
```

**Comment by agentpoyo** [comment]
> ⬆️ @wavell @denverwilliams take a look and merge if you 👍🏻

@taylor here are my findings: https://github.com/cnti-testcatalog/testsuite/pull/1964#pullrequestreview-2013146572

**Comment by barmull** [comment]
Everything from the output is behaving expected. In this PR: https://github.com/cnti-testcatalog/testsuite/pull/1995 is an improvement to catch one condition:
- setting KUBECONFIG into existing file, but the file is empty which is resulting that KUBECONFIG is set, but cluster is not running

**Comment by agentpoyo** [comment]
> Everything from the output is behaving expected. In this PR: #1995 is an improvement to catch one condition:
> 
> * setting KUBECONFIG into existing file, but the file is empty which is resulting that KUBECONFIG is set, but cluster is not running

The output when the kubeconfig file configured is empty is too generic, it doesn't detail the exact issue.

```
Setup failed. Some prerequisites are missing. Please install all of the prerequisites before continuing.
```

**Comment by barmull** [comment]
> > Everything from the output is behaving expected. In this PR: #1995 is an improvement to catch one condition:
> > 
> > * setting KUBECONFIG into existing file, but the file is empty which is resulting that KUBECONFIG is set, but cluster is not running
> 
> The output when the kubeconfig file configured is empty is too generic, it doesn't detail the exact issue.
> 
> ```
> Setup failed. Some prerequisites are missing. Please install all of the prerequisites before continuing.
> ```

This error message is not part of ensure_kubeconfig method, probably there are some missing prerequisites for setup.

I tried to set KUBECONFIG to existing empty file and got this output (this behavior is expected):
image

It fails on this error message:  'Could not create cnf-testsuite namespace on the Kubernetes cluster' --> which means that cluster is not healthy. That's why I added an improvement that checks if cluster is up and running. (PR:#1995)

**Comment by martin-mat** [comment]
Looks good to me accordingly to intentions and scope of this PR.

**Comment by taylor** [comment]
@barmull @martin-mat what @agentpoyo was trying to do was do a QA to see if the binary was acting the way it was intending.  He was saying the output was not as expected.  Having Acceptance Crietia would help with testing.

This PR references the bug https://github.com/cnti-testcatalog/testsuite/issues/985.  The stack trace is not suppressed in the instances tested which was what the bug was about.  




----
Example of acceptance criteria, which should probably go into the issue and the PR should pass the acceptance criteria.


###  Common steps
- [x]  Find or create a K8s cluster to use
   - Talk with team for access to an existing cluster
   - [Kind K8s clusters](https://github.com/cncf/cnf-testsuite/blob/main/KIND-INSTALL.md) will work as well
- [x]  Set the KUBECONFIG environment variable for accessing the target K8s cluster
- [ ]  As a developer, Pull down latest version of CNF conformance test code from {FEATURE} branch
     - [ ]  Check that {FEATURE} Branch passes all A/C 
     - [ ]  Check that {FEATURE} Branch passes tests in GitHub Actions
- [x]  As a peer reviewer, check that PR passes in GitHub Actions
     - [x] if yes, merge PR to master
     - [ ]  if no, make a comment and request that developer make changes to pass tests 
- [x]   As a peer reviewer, Pull down latest version of CNF conformance test code from `main` branch
- [x]  Run `shards install`

### Update Prerequisite checker
- [ ] NA

### Update yml validator, if needed
- [ ]  update if needed

### I would expect Falco to successfully install when the falco setup command runs:
- [x] Falco is not installed on the K8s cluster or system
- [x] I run the setup command to install falco. Ex. `cnf-testsuite install_falco`
- [x] I should see a message saying Falco was successfully installed
- [x] I run the setup command to uninstall falco. Ex. `cnf-testsuite uninstall_falco`
- [x] I should see a message saying Falco was successfully uninstalled

### I would expect to PASSED for non_root_user test when running against a CNF which does not run processes as root:
- [x] Tell the test suite to use a CNF which does not run any processes as root. eg. `./cnf-testsuite cnf_setup cnf-path=sample-cnfs/sample_nonroot`
- [x]  Run the non-root user test: `./cnf-testsuite non_root_user`
- [x] I see a PASSED message. Example: PASSED: Root user not found 🚫√
- [x] I run the setup command to uninstall falco. Ex. `cnf-testsuite uninstall_falco`

### I would expect to see a FAILED result for the non_root_user test, when running against CNF which has processes running as the root :
- [x] Tell the test suite to use a CNF runs processes as root. eg. `./cnf-testsuite cnf_setup cnf-path=sample-cnfs/sample_coredns`
- [x]  Run the non-root user test: `./cnf-testsuite non_root_user`
- [x] I see a PASSED message. Example:  ❌ FAILED: Root user found 🚫√
- [x] I run the setup command to uninstall falco. Ex. `cnf-testsuite uninstall_falco`

**Comment by agentpoyo** [comment]
> Looks good to me accordingly to intentions and scope of this PR.

The stack trace after the error message should be suppressed. End users should not see crystal stack traces if cnf_setup fails to connect to a cluster. Also note, this is also caught during setup if a cluster is not reachable. 

Users who previously setup the testsuite could lose their config and without this code it defaults to say it could not create the namespace. A simple update to the display error detailing to "check your kubeconfig or cluster connection" is likely sufficient enough while suppressing the stack trace output that the original bug was pointing out.

This code does not suppress the stack trace from output which was the original intent of the bug.

**Comment by martin-mat** [comment]
> The stack trace after the error message should be suppressed. End users should not see crystal stack traces if cnf_setup fails to connect to a cluster. Also note, this is also caught during setup if a cluster is not reachable.

OK, this is a good point and stack trace should not be printed out to console output (perhaps only to an error log) 

> 
> Users who previously setup the testsuite could lose their config and without this code it defaults to say it could not create the namespace. A simple update to the display error detailing to "check your kubeconfig or cluster connection" is likely sufficient enough while suppressing the stack trace output that the original bug was pointing out.

This is not scope of bug report #985. As already stated in previous comments, a new ticket was written for it (#1968) and a different pull request addresses it (#1995)

**Comment by martin-mat** [comment]
> @martin-mat if the scope of the PR is not supposed to cover the bug that was referenced it should probably reference a new issue for what is in this PR. The acceptance criteria should be added for the intention of the PR.

this PR is intended to address the bug report. I think there was not obvious to me that @agentpoyo addresses by his test report displaying backtrace to the console below the error message. That should be removed, I agree.

**Comment by agentpoyo** [comment]
> This is not scope of bug report #985. As already stated in previous comments, a new ticket was written for it (#1968) and a different pull request addresses it (#1995)

Yes, I was just mentioning it since this one and the other kind of go hand in hand. Regardless, this PR does not address the stack trace output and the new one created is kind of redundant since we already had checks in place to detail when a cluster is not reachable in a sense.

Cheers.

**Comment by barmull** [comment]
Did the change. It should be without back trace now.

This is output when KUBECONFIG is unset and default path does not exist:
1unset_no default config path

This is output when KUBECONFIG is set to non-existing file:
2non-exiting file

**Comment by martin-mat** [comment]
@agentpoyo please re-review. The stacktrace removed.

**Comment by martin-mat** [comment]
gh actions passed
https://github.com/cnti-testcatalog/testsuite/actions/runs/9588046938

**Comment by martin-mat** [comment]
Prior comments addressed. Merging.

**Pull request merged**